src/headerChainSync/HeaderChainProvider.js

```diff
@@ -0,0 +1,205 @@
+/**
+ * This module syncs the header chain
+ * No need to use this module manually - it's part of HeaderChainSync.
+ * @module HeaderChainProvider
+ */
+const { SpvChain } = require('@dashevo/dash-spv');
+const Api = require('../');
+const MNDiscovery = require('../MNDiscovery/index');
+const config = require('../config');
+const constants = require('../constants');
+
+/**
+ * This class syncs the header chain in a parallel manner across masternodes
+ */
+class HeaderChainProvider {
+  /**
+   * @param {string} network
+   * @param {SimplifiedMNListEntry[]} seeds
+   * @param {number} DAPIPort
+   */
+  constructor(network, seeds, DAPIPort = config.Api.port) {
+    this.network = network;
+
+    const seedsIsArray = Array.isArray(seeds);
+
+    if (seeds && !seedsIsArray) {
+      throw new Error('seed is not an array');
+    }
+
+    this.seeds = seedsIsArray ? seeds.slice() : config.DAPIDNSSeeds.slice();
+    this.DAPIPort = DAPIPort;
+    this.MNDiscovery = new MNDiscovery(this.seeds, this.DAPIPort);
+  }
+
+  /**
+   * @private
+   *
+   * Retrieve headers for a slice and populate header chain
+   *
+   * @param {DAPIClient} api
+   * @param {SpvChain} headerChain
+   * @param {object} options
+   * @param {int} options.fromHeight
+   * @param {int} options.toHeight
+   * @param {int} options.step
+   * @param {int} options.offset
+   * @param {int} [options.retryCount=0]
+   * @param {int} [options.extraHeight=0]
+   *
+   * @returns {Promise<void>}
+   */
+  async populateHeaderChain(
+    api, headerChain, options,
+  ) {
+    const {
+      fromHeight, toHeight, step, offset, retryCount = 0, extraHeight = 0,
+    } = options;
+
+    for (let height = fromHeight; height < toHeight - extraHeight; height += step) {
+      /* eslint-disable-next-line no-await-in-loop */
+      const newHeaders = await api.getBlockHeaders(height, step);
+      try {
+        headerChain.addHeaders(newHeaders);
+      } catch (e) {
+        if (retryCount > 0) {
+          /* eslint-disable-next-line no-await-in-loop */
+          await this.populateHeaderChain(
+            api, headerChain, {
+              fromHeight, toHeight, step, offset, retryCount: retryCount - 1,
+            },
+          );
+        }
+      }
+    }
+
+    if (extraHeight > 0) {
+      const extraHeaders = await api.getBlockHeaders(toHeight, extraHeight);
+      try {
+        headerChain.addHeaders(extraHeaders);
+      } catch (e) {
+        if (retryCount > 0) {
+          await this.populateHeaderChain(
+            api, headerChain,
+            {
```

Comment:
Probably it's better to have Chunk object with state instead of passing options.